src/core/net/srp_client.cpp

```diff
@@ -1428,6 +1460,75 @@ void Client::HandleTimer(void)
     }
 }
 
+#if OPENTHREAD_CONFIG_SRP_CLIENT_AUTO_START_API_ENABLE
+
+void Client::EnableAutoStartMode(AutoStartCallback aCallback, void *aContext)
+{
+    mAutoStartCallback = aCallback;
+    mAutoStartContext  = aContext;
+
+    VerifyOrExit(!mAutoStartModeEnabled);
+    mAutoStartModeEnabled = true;
+    CheckAutoStart();
+
+exit:
+    return;
+}
+
+void Client::CheckAutoStart(void)
+{
+    uint16_t                                numServers = 0;
+    NetworkData::Service::SrpServer::Info   selectedServer;
+    NetworkData::Service::SrpServer::Info   server;
+    NetworkData::Service::Manager::Iterator iterator;
+
+    VerifyOrExit(mAutoStartModeEnabled);
+
+    // If the client is not running we check if there is any SRP sever
+    // info in Network Data and select one randomly and then start the
+    // client. If the client is already running with a server that was
+    // selected by the auto-start feature, we verify that the selected
+    // server is still present in the Network Data.
+
+    VerifyOrExit(!IsRunning() || mAutoStartDidSelectServer);
```

Comment:
Just curious when could it happen that `IsRunning() && !mAutoStartDidSelectServer` can be satisfied. 
I think it should be that the user started the SRP client manually and then enables `auto start` mode when the SRP client is already running. 
This seems to be a mixed use of `auto start` and `manual` modes, which could make things more complex than necessary. 

I think it would be slightly better if there is a distinct boundary between the two modes:
- `auto start` mode: always use auto selected server from network data
- `manual` mode: always use the manually specified server 

That would being following restrictions to the user APIs:
- User can not enable `auto start` mode when the SRP client is already running (manually started). User must stop the manually started SRP first. 
- User can not manually start the SRP server when `auto start` mode is enabled, even when there is no SRP server TLV in the network data (i.e. SRP client is not running). If I understand correctly, this is allowed by this PR which I think is quite a confusing behavior. 

Thoughts?

Reply:
> when could it happen that `IsRunning() && !mAutoStartDidSelectServer` can be satisfied.

You mention one way, but the more common way is that auto-start is enabled by default (which is the feature we support through an OT config) and then explicit call to `Start()` client with a specific server. 

---

The API policy/behavior we have is as follows:
- `Start()` can be called to start the client if it is not already running.
- Auto-start mode can be enabled/disabled at any time.
- When client is started by an explicit call to `Start()` it continues to use SRP server specified by `Start()`: 
   - (i.e., we won't check the SRP server to be present in list of servers discovered by auto-start feature).

I feel this is a good and simple set of API policies.

We can add more conditions/restrictions that for example require user to disable auto-start mode before calling `Start()` and/or not allow user to change auto-start mode after `Start()` but there are some downsides to this:
- We need to add more code to check for these combinations and return error to user (increase code size for not so common behaviors).
- It makes the documentation/description of the APIs (the API use policy) more complex.
- It's true that user can use the APIs in uncommon ways (like enable `auto-start` after calling `Start()` which is pointless) but the way I see it is that as long as we behave consistently and correctly, we don't need to explicitly protect user from using uncommon API combinations. 

-------

For longer-term, we may add use-cases where we expand the auto-start feature and want to use/enable it even after the client is started by `Start()` call. 
   - A possible future enehancment is to use auto-start mode to (also) help switch the SRP server if we fail to get a response from the currently selected SRP server after many attempts/retries. 
      - In the current implementation when we select a server, we stick with it even if we never receive a response from it. We switch the server when the SRP server entry/info is not longer seen in Thread network-data.
      - Basically the idea of this enhancement is that if we don't get a response from the current SRP server (after many retries), we try to see if there is any other SRP server options (from network-data) and switch to it.
   - We may want to use/enable this behavior for both cases where the SRP server is selected by auto-start or through `Start()` call.

Reply:
> > when could it happen that `IsRunning() && !mAutoStartDidSelectServer` can be satisfied.
> 
> You mention one way, but the more common way is that auto-start is enabled by default (which is the feature we support through an OT config) and then explicit call to `Start()` client with a specific server.

Do you mean a explicit call to `Start()` with one server chosen from the Network Data, meaning the user want to implement custom rules for choosing a server?

> We can add more conditions/restrictions that for example require user to disable auto-start mode before calling `Start()` and/or not allow user to change auto-start mode after `Start()` but there are some downsides to this:
> 
> * We need to add more code to check for these combinations and return error to user (increase code size for not so common behaviors).

Not necessarily to increase code size? Currently we have code logic checking whether the server is actually chosen by `auto-start` feature even when `auto-start` feature is ON. These logic won't be necessary in the proposed strategy. 

> * It makes the documentation/description of the APIs (the API use policy) more complex.
> * It's true that user can use the APIs in uncommon ways (like enable `auto-start` after calling `Start()` which is pointless) but the way I see it is that as long as we behave consistently and correctly, we don't need to explicitly protect user from using uncommon API combinations.

In this current strategy, enabling `auto-start` after calling `Start()` would return OK. I think this might be a little bit confusing to users. It seems that `auto-start` is successfully, but the client is still **NOT** using servers in Network Data.

> For longer-term, we may add use-cases where we expand the auto-start feature and want to use/enable it even after the client is started by `Start()` call.
> 
> * A possible future enehancment is to use auto-start mode to (also) help switch the SRP server if we fail to get a response from the currently selected SRP server after many attempts/retries.
>   
>   * In the current implementation when we select a server, we stick with it even if we never receive a response from it. We switch the server when the SRP server entry/info is no longer seen in Thread network-data.
>   * Basically the idea of this enhancement is that if we don't get a response from the current SRP server (after many retries), we try to see if there is any other SRP server options (from network-data) and switch to it.
> * We may want to use/enable this behavior for both cases where the SRP server is selected by auto-start or through `Start()` call.

I think this is actually a good point. I think the `auto-start` feature can simply be a automatic way of adding servers, but the strategy of selecting server is independent on whether the servers are input from Network Data or user configuration. User would also be able to add multiple servers for the SRP client to chose the best server. 

That being said, I am fine that we work out a better solution in the future.